### DISCUSSION TITLE
how can I do array_append and array_remove on supabase js client?

### DISCUSSION THREAD

**Discussion by Cleysonlb** [supabase/supabase#2771] [Questions] [answered]
Hi guys, I tried do something like `array_append` and `array_remove` but do not works...
How can I do this commands inside the supabase JS client?
Thanks

**Comment by dshukertjr** [comment, marked as answer]
Hi @Cleysonlb!

For now, you can write stored procedures like this to achieve do `array_append` and `array_remove`. 

```sql
create table if not exists public.array_table (
    id uuid not null primary key default uuid_generate_v4 (),
    array_element text[] not null default array[]::text[]
);

create or replace function append_array(new_element text, id uuid)
returns void
as
$func$
    update array_table
    set array_element = array_append(array_element, $1)
    where id = $2
$func$
language sql;

create or replace function remove_array(new_element text, id uuid)
returns void
as
$func$
    update array_table
    set array_element = array_remove(array_element, $1)
    where id = $2
$func$
language sql;
```

```typescript

let { data, error } = await supabase
  .rpc('append_array', {
    id, 
    new_element
  })
```

```typescript
let { data, error } = await supabase
  .rpc('remove_array', {
    id, 
    new_element
  })
```

By the way, what are you storing in those arrays?

**Reply by Cleysonlb** [reply]
Hi @dshukertjr, thanks for the answer, I'm storing ID's on this array, and sometimes I need "move" some ID for example, I have a `waiting_approval` table and `approved` table, when the ID is approved I need to move this ID from `waiting_approval` to `approved`

I will try your solution :+1:

**Reply by dshukertjr** [reply]
I presume you come from a Firebase background like I am (sorry if you're not 😂). We tend to store id's as arrays in one of the columns, because that is how we did things in Firebase, but generally, that is not a good practice in SQL. You should create a separate relational table.

**Reply by webdesignberlin** [reply]
is it possible to prevent duplicate entries from being added?

**Reply by dshukertjr** [reply]
@webdesignberlin You could do something like [this](https://dba.stackexchange.com/questions/91209/append-elements-to-array-only-if-the-element-doest-exist) to add an element to an array only if it does not exist, but again in SQL, you generally want to store array element in a separate table for better data quality.

**Reply by Thenlie** [reply]
This is a great solution, thank you! Is there a simple way to return the updated row rather than void?

**Comment by davilico05** [comment]
Hi! How I handle this with RLS policies? Because I have an update rule
`CREATE POLICY "Enable insert for authenticated users only" ON public.comments FOR UPDATE WITH CHECK (auth.role() = 'authenticated');`

**Reply by dshukertjr** [reply]
Hi @davilico05!

If you could open a new discussion with your specific question, I or someone else should be able to answer it!

**Reply by davilico05** [reply]
Done! [#3571](https://github.com/supabase/supabase/discussions/3571)
